backend/chunker/chunker.go

```diff
@@ -0,0 +1,1412 @@
+	err := configstruct.Set(m, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	remote := opt.Remote
+	if strings.HasPrefix(remote, name+":") {
+		return nil, errors.New("can't point remote at itself - check the value of the remote setting")
+	}
+
+	wInfo, wName, wPath, wConfig, err := fs.ConfigFs(remote)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse remote %q to wrap", remote)
+	}
+	// Look for a file first
+	remotePath := fspath.JoinRootPath(wPath, rpath)
+	wrappedFs, err := wInfo.NewFs(wName, remotePath, wConfig)
+	if err != fs.ErrorIsFile && err != nil {
+		return nil, errors.Wrapf(err, "failed to make remote %s:%q to wrap", wName, remotePath)
+	}
+
+	f := &Fs{
+		Fs:   wrappedFs,
+		name: name,
+		root: rpath,
+		opt:  *opt,
+	}
+
+	switch opt.MetaFormat {
+	case "none":
+		f.useMeta = false
+	case "simplejson":
+		f.useMeta = true
+	default:
+		return nil, fmt.Errorf("unsupported meta format '%s'", opt.MetaFormat)
+	}
+
+	if err := f.parseNameFormat(opt.NameFormat); err != nil {
+		return nil, fmt.Errorf("invalid name format '%s': %v", opt.NameFormat, err)
+	}
+
+	// Note 1: the features here are ones we could support, and they are
+	// ANDed with the ones from wrappedFs.
+	// Note 2: features.Fill() points features.PutStream to our PutStream,
+	// but features.Mask() will nullify it if wrappedFs does not have it.
+	f.features = (&fs.Features{
+		CaseInsensitive:         true,
+		DuplicateFiles:          true,
+		ReadMimeType:            true,
+		WriteMimeType:           true,
+		BucketBased:             true,
+		CanHaveEmptyDirectories: true,
+		SetTier:                 true,
+		GetTier:                 true,
+	}).Fill(f).Mask(wrappedFs).WrapsFs(f, wrappedFs)
+
+	return f, err
+}
+
+// Options defines the configuration for this backend
+type Options struct {
+	Remote          string        `config:"remote"`
+	ChunkSize       fs.SizeSuffix `config:"chunk_size"`
+	NameFormat      string        `config:"name_format"`
+	StartFrom       int           `config:"start_from"`
+	MetaFormat      string        `config:"meta_format"`
+	FailOnBadChunks bool          `config:"fail_on_bad_chunks"`
+}
+
+// Fs represents a wrapped fs.Fs
+type Fs struct {
+	fs.Fs
+	wrapper    fs.Fs
+	name       string
+	root       string
+	useMeta    bool
+	nameFormat string
+	nameRegex  *regexp.Regexp
+	opt        Options
+	features   *fs.Features // optional features
+}
+
+// Name of the remote (as passed into NewFs)
+func (f *Fs) Name() string {
+	return f.name
+}
+
+// Root of the remote (as passed into NewFs)
+func (f *Fs) Root() string {
+	return f.root
+}
+
+// Features returns the optional features of this Fs
+func (f *Fs) Features() *fs.Features {
+	return f.features
+}
+
+// String returns a description of the FS
+func (f *Fs) String() string {
+	return fmt.Sprintf("Chunked '%s:%s'", f.name, f.root)
+}
+
+// parseNameFormat converts pattern-based name format into Printf format and Regexp
+func (f *Fs) parseNameFormat(pattern string) error {
```

Comment:
This should probably have a unit test which makes patterns and checks they perform as expected!